### PR TITLE
Spi

### PR DESCRIPTION
**Pull request by hmvp**
Basic implementation based on C SDK code. Does not seem to work yet

### CODE REVIEW

File: rp2040-hal/src/spi.rs [hunk 1]
```diff
@@ -1,3 +1,184 @@
...
+            }
+        }
+
+        impl Spi for $name {
+            fn init(&mut self, resets: &mut RESETS, baudrate: u32) {
+                self.spi_dev.reset_bring_down(resets);
+                self.spi_dev.reset_bring_up(resets);
+
+                self.set_baudrate(baudrate);
+                self.spi_dev
```
Comment (anall) [inline rp2040-hal/src/spi.rs:72]: This and the following write to `sspcsr0` clobber the result of `set_baudrate`

File: rp2040-hal/src/spi.rs [hunk 2]
```diff
@@ -1,3 +1,184 @@
...
+                        break;
+                    }
+                }
+
+                // We might not find a prescale value that lowers the clock freq enough, so we leave it at max
+                debug_assert_ne!(prescale, u8::MAX);
+
+                // Find largest post-divide which makes output <= baudrate. Post-divide is
+                // an integer in the range 1 to 256 inclusive.
+                for postdiv_option in 256u32..1 {
```
Comment (anall) [inline rp2040-hal/src/spi.rs:107]: `256u32..1` is an empty range, you need to do something like `(1..=256u32).rev()`

**Review by anall — COMMENTED** [review]
Just adding some notes for your future reference. Those changes + some extra support for additional `embedded_hal` traits got the OLED screen on the test board (Adafruit Macropad) working

File: rp2040-hal/src/spi.rs (hunk 1 above)
Comment (hmvp) [reply rp2040-hal/src/spi.rs]: Ah yes some debugging code.. Should be fixed now

File: rp2040-hal/src/spi.rs (hunk 2 above)
Comment (hmvp) [reply rp2040-hal/src/spi.rs]: Fixed

File: rp2040-hal/src/spi.rs [hunk 3]
```diff
@@ -1,3 +1,218 @@
...
+}
+
+impl<D: SpiDevice, const DS: u8> Spi<Enabled, D, DS> {
+    fn is_writable(&self) -> bool {
+        self.device.sspsr.read().tnf().bit_is_set()
+    }
+    fn is_readable(&self) -> bool {
+        self.device.sspsr.read().rne().bit_is_set()
+    }
+    // fn is_busy(&self) -> bool {
```
Comment (9names) [inline rp2040-hal/src/spi.rs:167]: Commented code should have a comment to explain why it is commented out

**Review by 9names — CHANGES_REQUESTED** [review]
Pretty good port of the C SDK code, just a couple of things I'd like cleaned up.

If you wouldn't mind, could you please not rewrite git history after a review?
I can squash the PR into a single commit on merge if you like, but it's hard to follow review history and review-related fixes when they aren't separate commits on top of your existing work.

File: rp2040-hal/src/spi.rs [hunk 4]
```diff
@@ -1,3 +1,218 @@
...
+    pub fn new(device: D) -> Spi<Disabled, D, DS> {
+        Spi {
+            device,
+            state: PhantomData,
+        }
+    }
+    /// Set baudrate based on system clock
+    ///
+    /// Typically the systemclock is set to 125_000_000
+    fn set_baudrate(&mut self, sys_freqency: u32, baudrate: u32) -> u32 {
```
[the remark this answers is not in the document]
Comment (anall) [reply rp2040-hal/src/spi.rs:83]: This is based on the peripheral clock, not the system clock. By default, they are the same but do not need to be. On my test board, I'm running the system clock at 125MHz but have the peripheral clock running at 48MHz based on the USB PLL.

[the remark this answers is not in the document]
Comment (hmvp) [reply rp2040-hal/src/spi.rs:83]: Ah yes I am not entirely happy about that part. I feel the clocks struct should keep track of the desired frequency of all clocks so users can just ask for the right frequency and the code can figure out the dividers. But that needs a lot more work in the clocks section.

For now I will fix the comment and the variable name

File: rp2040-hal/src/spi.rs [hunk 5]
```diff
@@ -1,3 +1,214 @@
...
+    pub fn new(device: D) -> Spi<Disabled, D, DS> {
+        Spi {
+            device,
+            state: PhantomData,
+        }
+    }
+    /// Set baudrate based on peripheral clock
+    ///
+    /// Typically the peripheral clock is set to 125_000_000
+    fn set_baudrate(&mut self, peri_frequency: u32, baudrate: u32) -> u32 {
```
Comment (hmvp) [inline rp2040-hal/src/spi.rs:82]: Should I use embedded_time::rate::*Hertz here for the frequency?

Comment (anall) [reply rp2040-hal/src/spi.rs:82]: That would make this easier to use, at least.

**Comment by anall** [comment]
I'll look at this more after work today, but this works fine in my (Adafruit Macropad) test program with the only changes required being to support the different calling conventions.

I wonder if we want to add the ability to set the frame format too?

**Review by 9names — APPROVED** [review]

**Comment by 9names** [comment]
I think we will want different frame formats, but we can wait until we have a need for them before we implement them.
I'll wait for @anall to do additional testing before merging this.

**Comment by anall** [comment]
No need to hold on my review as this does work on my test board and we can
iterate later.

(Also: using the TI frame format does make the timing a little less gross
when (mis-)using SPI to drive WS2812 LEDS/"Neopixel"s, but that can be a
future improvement)

**Comment by 9names** [comment]
Great, merging as-is then.
Thanks y'all!
